contributors/devel/api_changes.md

```diff
@@ -943,30 +945,24 @@ One possible place to do this is in the REST storage strategy's PrepareForCreate
       newFrobber := obj.(*api.Frobber)
       oldFrobber := old.(*api.Frobber)
     
-      if !utilfeature.DefaultFeatureGate.Enabled(features.Frobber2D) {
+      if !utilfeature.DefaultFeatureGate.Enabled(features.Frobber2D) && oldFrobber.Width == nil {
         newFrobber.Width = nil
-        oldFrobber.Width = nil
       }
     }
     ```
 
-4. In validation, ensure the alpha field is not set if the feature gate is disabled (covers cases we might miss in the above):
+4. In validation, validate the field if present:
```

Comment:
I'm not sure I agree with this.  Philosophically, is a gated-off feature supposed to behave like it doesn't exist?  Or is it supposed to be detectable?


E.g.

Consider a user today who has YAML that specifies a non-existent field.  Maybe they mis-typed it, or mis-indented it or something.  Regardless, it doesn't map to a real field in API, and they haven't noticed that it isn't doing anything.  This seems utterly plausible to me.

The API today will allow it and just drop the unknown field.

Now we add an alpha feature, with a gate, and we disable the gate.  This feature happens to introduce a field with the name the user used.  Not impossible.  Upgrade the user's cluster.

When the YAML is sent to the upgraded cluster, should it suddenly start failing?  The feature is alpha and gated-off.  Why should it have any effect at all?

What about a case where a user has stored state, which you go out of your way to preserve above, which would fail validation?  Even with the feature gated off, does that have impact?  I'm not sure, I guess validation doesn't run at that point, so maybe not.

Reply:
> is a gated-off feature supposed to behave like it doesn't exist? Or is it supposed to be detectable?

It is supposed to behave like it doesn't exist, *as much as possible*. Existing persisted data for a disabled field is problematic. Simply dropping the field is not uniformly safe and can result in an impossible-to-update (and therefore impossible-to-remove) object.

> Consider a user today who has YAML that specifies a non-existent field. Maybe they mis-typed it, or mis-indented it or something. Regardless, it doesn't map to a real field in API, and they haven't noticed that it isn't doing anything. This seems utterly plausible to me.
> 
> The API today will allow it and just drop the unknown field.

Correct. This is not proposing changing that behavior. Persisted objects in etcd do not contain typoed non-existent fields.

> Now we add an alpha feature, with a gate, and we disable the gate. This feature happens to introduce a field with the name the user used. Not impossible. Upgrade the user's cluster.
>
> When the YAML is sent to the upgraded cluster, should it suddenly start failing? The feature is alpha and gated-off. Why should it have any effect at all?

It would not start failing. PrepareForCreate would drop the field, because the feature was disabled. When validation was called, the field would be empty, and validation would not fail.

> What about a case where a user has stored state, which you go out of your way to preserve above, which would fail validation? Even with the feature gated off, does that have impact? I'm not sure, I guess validation doesn't run at that point, so maybe not.

Data that is persisted has already passed validation. We are constrained from allowing data to be persisted that the previous version of the API server would fail validation on.